src/node_data.js

```diff
@@ -94,21 +94,50 @@ const initData = function(node, nodeName, key) {
 /**
  * Retrieves the NodeData object for a Node, creating it if necessary.
  *
- * @param {Node} node The node to retrieve the data for.
+ * @param {?Node} node The Node to retrieve the data for.
  * @return {!NodeData} The NodeData for this Node.
  */
 const getData = function(node) {
   let data = node['__incrementalDOMData'];
 
   if (!data) {
-    const nodeName = node.nodeName.toLowerCase();
-    let key = null;
+    data = importNode(node);
+  }
+
+  return data;
+};
+
+
+/**
+ * Imports node and its subtree, initializing caches.
+ *
+ * @param {?Node} node The Node to import.
+ * @return {!NodeData} The NodeData for this Node.
+ */
+const importNode = function(node) {
+  const nodeName = node.nodeName.toLowerCase();
+  const isElement = node instanceof Element;
+  const key = isElement ? node.getAttribute('key') : null;
+  const data = initData(node, nodeName, key);
+
+  if (key) {
+    getData(node.parentNode).keyMap[key] = node;
+  }
 
-    if (node instanceof Element) {
-      key = node.getAttribute('key');
+  if (isElement) {
+    for (let i = 0; i < node.attributes.length; i += 1) {
```

Comment:
I kind of assume that JavaScript engines can cache these lookups, but changed it just to be safe.